### ISSUE TITLE
Devise shows misleading "reset password token is invalid" error when including authentication key in strip_whitespace_keys

### ISSUE THREAD

**Issue by rockorequin** [plataformatec/devise#3431]
With devise 3.4.1 (and also the latest from git as of 2015-01-19) and Rails 4.1.8, if I try to reset my password I get a misleading error message "reset password token is invalid" if I have this in the config/initializers/devise.rb file:

<code>
config.authentication_keys = [ :user_name ]
</code>
<code>
config.strip_whitespace_keys = [ :user_name ]
</code>

I have stepped through the code and found that the encrypted token that Devise creates from the posted token matches the reset_password_token in the user record exactly, so "reset password token is invalid" is not actually the issue.

Instead, what happens is that lib/devise/models/recoverable.rb#reset_password_by_token has this line:

<code>
recoverable = find_or_initialize_with_error_by(:reset_password_token, reset_password_token)
</code>

and this method eventually calls lib/devise/parameter_filter.rb#filter to construct a conditions hash to find the user record using the reset_password_token that was calculated. However, this line in the filter method:

<code>
conditions.merge!(filtered_hash_by_method_for_given_keys(conditions.dup, :strip, @strip_whitespace_keys))
</code>

adds { user_name: nil } to the conditions hash, and as a result the find first method returns no user record (obviously since no records have a null authentication field in the database). Since the user record is null, reset_password_by_token then creates a new user record, which has the "reset password token is invalid" error in it.

The workaround is to remove the authentication key :user_name from the strip_whitespace_keys, and the error no longer occurs.

It might also be relevant that I have configured Devise to use the Person model (via "devise_for :people..." in config/routes.rb).

Am I misusing the strip_whitespace_keys configuration? I don't see why devise insists on adding user_name to the conditions hash when it hasn't been passed user_name as an attribute to match.

**Comment by josevalim** [comment]
Thanks for the detailed description. You are right, this is a bug, can you please send a pull request?

**Comment by rockorequin** [comment]
No worries. I don't actually have a patch for this, though - I simply changed my configuration to avoid the issue, which is only a workaround. I don't see why the filter method is including the strip_whitespace_keys attributes, so if we were to remove it, I assume it might break something else.

**Comment by josevalim** [comment]
>  I don't see why the filter method is including the strip_whitespace_keys attributes, so if we were to remove it, I assume it might break something else.

Yeah, this looks like a bug. It should not be included definitely.

**Comment by sahilbhatia** [comment]
@rockorequin 

I tried reproducing it (from the details given above) on my local machine but couldn't.

I used ruby ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-linux] along with the rails and devise versions you specified.

I made one additional change in `config/initializers/devise.rb`
 i.e. added this line `config.reset_password_keys = [ :username ]`

Am I missing anything or going wrong somewhere?

**Comment by rockorequin** [comment]
@sahilbhatia, I'm using ruby 2.2.0p0 as well (and Rails 4.0.9). I think these are the only changes I made to the standard config file (as of 26 Aug 2013):

config.authentication_keys = [ :user_name ]
config.strip_whitespace_keys = [ :user_name ]
config.reset_password_keys = [ :user_name ]
config.case_insensitive_keys = [ :email ]
config.lock_strategy = :failed_attempts
config.unlock_keys = [ :email ]
config.maximum_attempts = 5

Note that I'm using user_name and not username because that is the field's name in the database (and of course I'm using the People model instead of the User model). Could that be making a difference?

**Comment by andricicezar** [comment]
I have same problem. I get the error: "Reset password token is invalid".
Rails 4.2.1
Ruby 2.1.2p95
Devise 3.5.1

From header, form data:
```
utf8:✓
_method:put
authenticity_token:rhrOrwfqQiO4oLaQHD5/yxQfmBGqDlhIC6NtCLDYR+ewt/pgZRSTx80x6SANnNRodjwp+x6wQ9/whSzNdYMqog==
user[reset_password_token]:3251df0b175ee589f2f8244eb40f9261b7c6cf7126b57e946aec33d365a253c9
user[password]: ***
user[password_confirmation]: ***
commit:Change my password
```

In logs:
```
Started PUT "/users/password" for 127.0.0.1 at 2015-06-12 11:24:02 +0300
Processing by Devise::PasswordsController#update as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"rhrOrwfqQiO4oLaQHD5/yxQfmBGqDlhIC6NtCLDYR+ewt/pgZRSTx80x6SANnNRodjwp+x6wQ9/whSzNdYMqog==", "user"=>{"reset_password_token"=>"[FILTERED]", "password"=>"[FILTERED]", "password_confirmation"=>"[FILTERED]"}, "commit"=>"Change my password"}
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE "users"."reset_password_token" = ?  ORDER BY "users"."id" ASC LIMIT 1  [["reset_password_token", "14e8537a577e548a9c507854fac1d203fc5b2babff6334c4704295d5b127b832"]]
```

The reset_password_token is different in logs than in header.

I tried to comment the config.strip_whitespace_keys but didn't fix the bug.

**Comment by guilpejon** [comment]
Having the same problem as @andricicezar, with the same versions of Rails and Devise 3.4.1.

**Comment by andricicezar** [comment]
@guilpejon I fixed my problem with this. https://github.com/plataformatec/devise/issues/2699

**Comment by topherfangio** [comment]
@rockorequin Can you tell me if this was happening 100% of the time, or just sometimes? I am facing a similar issue in 3.5.2 and am trying to track down the exact issue; but it's not always happening.

**Comment by rockorequin** [comment]
@topherfangio: it was quite a while ago now since I ran into this, but I'm pretty certain it was happening 100% of the time.

**Comment by abnud1** [comment]
I'm facing this problem even when I don't have authentication keys in strip_whitespace_keys
using Rails 4.2.5 with devise 3.5.3 and ruby 2.2.3

**Comment by abnud1** [comment]
stepping through code shows that the encrypted token gets saved in the database, the mail sent contains a link in which the raw token is used , I have verified this by inspecting the message variable and message.body.  

However when I hover over the link in the mail preview in chromium 47 I see the link with the raw token but then the link changes and the encrypted token is exposed , I don't know why this happens and When I click the link, the reset_token in the url paramters is the encrypted one not the raw one even though inspecting the message variable shows the opposite of this.  

I Have verified that the same thing happens with Firefox developer edition, why is this happening ?

**Comment by abnud1** [comment]
Ok I have verfied, it was my fault, I was using the token from the database which is encrypted in the Mailer preview class.

**Comment by joshpencheon** [comment]
I was having this issue with Rails 5.0 and Devise 4.2.

I tracked it down to when `reset_password_by_token` tries to find the user to reset using the digested token. Like @rockorequin, I was seeing an additional predicate being added to the query (`email: nil`), meaning that the user was not found, and a stub was initialised instead to carry the "token is invalid" message back to the form.

Digging further, I found that it was `Devise::ParameterFilter` that was adding the additional `email:` key to the search. Finally, it turned out this was because another legacy library being used by the project defined `nil#downcase` and `nil#strip`.

In this case, not Devise's fault at all. For those interested, my patch was:

```diff
    def filtered_hash_by_method_for_given_keys(conditions, method, condition_keys)
      condition_keys.each do |k|
+       next unless conditions.key?(k)
+
        value = conditions[k]
        conditions[k] = value.send(method) if value.respond_to?(method)
      end

      conditions
    end   
```

I don't think it would be _wrong_ make this change upstream – whilst the cause in this case was horrible patching outside of Devise's control, I don't believe the intention of `filtered_hash_by_method_for_given_keys` is to add new keys to `conditions`, and think my patch makes that obvious.

Hope this helps someone 😃
